### PR TITLE
feat: allow user to configure manifest validation by setting deploy.kubectl.dryRun

### PR DESCRIPTION
**Pull request by MarlonGamez**
**Related**: #6273 

**Description**
This PR introduces a new config option that allows users to change how their manifests are validated.
Users can now configure `deploy.kubectl.dryRun` in their `skaffold.yaml`. Values can be `none`, `server`, or `client`.
This will simply change the value that is passed to `kubectl create --dry-run=`.

**Follow-up Work (remove if N/A)**
As per #6273, we should change the default value to `server`. I will do that in a follow up PR.

### CODE REVIEW

File: pkg/skaffold/deploy/kubectl/cli.go [hunk 1]
```diff
@@ -205,7 +207,12 @@ func (c *CLI) ReadManifests(ctx context.Context, manifests []string) (manifest.M
 		return nil, versionGetErr(err)
 	}
 	if compTo1_18 >= 0 {
-		dryRun += "=client"
+		// Always use `client` on cleanup to avoid errors from resources already existing on cluster
```
Comment (briandealwis) [inline pkg/skaffold/deploy/kubectl/cli.go:210]: This is unexpected.  Is there a pointer for more details?  Just some resources?

File: pkg/skaffold/deploy/kubectl/cli.go [hunk 2]
```diff
@@ -192,8 +194,8 @@ func (c *CLI) WaitForDeletions(ctx context.Context, out io.Writer, manifests man
 	}
 }
 
-// ReadManifests reads a list of manifests in yaml format.
-func (c *CLI) ReadManifests(ctx context.Context, manifests []string) (manifest.ManifestList, error) {
+// ReadManifests reads a list of manifests in yaml format and validates them by running `kubectl create --dry-run` .
+func (c *CLI) ReadManifests(ctx context.Context, manifests []string, cleanup bool) (manifest.ManifestList, error) {
```
Comment (briandealwis) [inline pkg/skaffold/deploy/kubectl/cli.go:198]: Could you document the meaning of cleanup too.

File: pkg/skaffold/deploy/kubectl/kubectl_test.go [hunk 3]
```diff
@@ -219,6 +219,56 @@ func TestKubectlDeploy(t *testing.T) {
 			shouldErr:        true,
 			waitForDeletions: true,
 		},
+		{
```
Comment (briandealwis) [inline pkg/skaffold/deploy/kubectl/kubectl_test.go:222]: Should include a test for `none` too

File: pkg/skaffold/deploy/kubectl/cli.go [hunk 4]
```diff
@@ -62,12 +63,13 @@ type Config interface {
 	PipelineForImage(imageName string) (latestV1.Pipeline, bool)
 }
 
-func NewCLI(cfg Config, flags latestV1.KubectlFlags, defaultNamespace string) CLI {
+func NewCLI(cfg Config, flags latestV1.KubectlFlags, defaultNamespace string, dryRun string) CLI {
```
Comment (briandealwis) [inline pkg/skaffold/deploy/kubectl/cli.go:66]: Add some doc here.  If dryRun == "" then uses "client".

File: pkg/skaffold/schema/defaults/defaults.go [hunk 5]
```diff
@@ -208,6 +209,12 @@ func setDefaultLogsConfig(c *latestV1.SkaffoldConfig) {
 	}
 }
 
+func setDefaultKubectlDryRun(c *latestV1.SkaffoldConfig) {
+	if c.Deploy.KubectlDeploy != nil && c.Deploy.KubectlDeploy.DryRun == "" {
+		c.Deploy.KubectlDeploy.DryRun = "client"
+	}
+}
```
Comment (briandealwis) [inline pkg/skaffold/schema/defaults/defaults.go:216]: Do we need this? You're already handling `""` in the CLI code itself, and it provides false comfort since you're also passing in `""` explicitly from the Kustomize code.

File: pkg/skaffold/schema/latest/v1/config.go [hunk 6]
```diff
@@ -568,6 +568,13 @@ type KubectlDeploy struct {
 	// DefaultNamespace is the default namespace passed to kubectl on deployment if no other override is given.
 	DefaultNamespace *string `yaml:"defaultNamespace,omitempty"`
 
+	// DryRun is the value kubectl will use with `kubectl create --dry-run=`. Valid values are:
+	// `none`: validate manifests using `kubectl create --dry-run=none`.
+	// `server`: validate manifests using `kubectl create --dry-run=server`.
+	// `client` (default): validate manifests using `kubectl create --dry-run=client`.
```
Comment (briandealwis) [inline pkg/skaffold/schema/latest/v1/config.go:574]: I think it makes sense to order the default first, and place `none` last?

File: pkg/skaffold/schema/latest/v1/config.go [hunk 7]
```diff
@@ -568,6 +568,13 @@ type KubectlDeploy struct {
 	// DefaultNamespace is the default namespace passed to kubectl on deployment if no other override is given.
 	DefaultNamespace *string `yaml:"defaultNamespace,omitempty"`
 
+	// DryRun is the value kubectl will use with `kubectl create --dry-run=`. Valid values are:
```
Comment (briandealwis) [inline pkg/skaffold/schema/latest/v1/config.go:571]: Would you mind fixing the schema enum collector to support an optional trailing `:`?
https://github.com/GoogleContainerTools/skaffold/blob/151b8d373d1eafcfe559fa1b302907c893cdda1a/hack/schemas/main.go#L52

File: docs/content/en/schemas/v2beta24.json [hunk 8]
```diff
@@ -2586,6 +2586,12 @@
           "description": "default namespace passed to kubectl on deployment if no other override is given.",
           "x-intellij-html-description": "default namespace passed to kubectl on deployment if no other override is given."
         },
+        "dryRun": {
+          "type": "string",
+          "description": "value kubectl will use with `kubectl create --dry-run=`. Valid values are: `none`: validate manifests using `kubectl create --dry-run=none`. `server`: validate manifests using `kubectl create --dry-run=server`. `client` (default): validate manifests using `kubectl create --dry-run=client`.",
+          "x-intellij-html-description": "value kubectl will use with <code>kubectl create --dry-run=</code>. Valid values are: <code>none</code>: validate manifests using <code>kubectl create --dry-run=none</code>. <code>server</code>: validate manifests using <code>kubectl create --dry-run=server</code>. <code>client</code> (default): validate manifests using <code>kubectl create --dry-run=client</code>.",
+          "default": "client"
```
Comment (briandealwis) [inline docs/content/en/schemas/v2beta24.json:2593]: Hmm the enum values should be showing up here.

**Review by briandealwis — APPROVED** [review]
A few nits

File: pkg/skaffold/schema/latest/v1/config.go (hunk 7 above)
Comment (briandealwis) [reply pkg/skaffold/schema/latest/v1/config.go:571]: Maybe we should do this separately as it looks like there are a bunch of similar enums that aren't picked up.

File: pkg/skaffold/deploy/kubectl/cli.go (hunk 1 above)
Comment (MarlonGamez) [reply pkg/skaffold/deploy/kubectl/cli.go:210]: [This stack overflow answer](https://stackoverflow.com/questions/47369351/kubectl-apply-vs-kubectl-create) helped me understand what I was seeing here a bit more.

Basically since we are using `kubectl create` to validate the manifests, in the case of us cleaning up, since we validate in this way, if we validate using `--dry-run=server` then we'll receive an error saying the resources already exist on cluster.

File: pkg/skaffold/deploy/kubectl/cli.go (hunk 4 above)
Comment (MarlonGamez) [reply pkg/skaffold/deploy/kubectl/cli.go:66]: Will do. I'll also move the empty string check to here rather than the ReadManifests function. That way it's already set to "client" in the case that the value ever gets read anywhere else.

File: pkg/skaffold/schema/defaults/defaults.go (hunk 5 above)
Comment (MarlonGamez) [reply pkg/skaffold/schema/defaults/defaults.go:216]: that's true, it doesn't look like this is needed. I'll remove this

File: pkg/skaffold/schema/latest/v1/config.go (hunk 7 above)
Comment (MarlonGamez) [reply pkg/skaffold/schema/latest/v1/config.go]: yup, I've fixed for this here for now by removing the `:`. I'll open up another PR to allow a trailing `:`

Comment (MarlonGamez) [reply pkg/skaffold/schema/latest/v1/config.go]: https://github.com/GoogleContainerTools/skaffold/pull/6651

File: pkg/skaffold/deploy/kubectl/cli.go (hunk 1 above)
Comment (briandealwis) [reply pkg/skaffold/deploy/kubectl/cli.go:217]: Oh! I didn't realize we ran this as `kubectl create`.  I wonder why.

**Comment by MarlonGamez** [comment]
Holding off on merging this change as we're awaiting more discussion on the topic.
CC @sivakku @briandealwis

**Comment by MarlonGamez** [comment]
closing this as we've decided that it's unneeded at the moment
